feat(content_safety): add support to auto select multilingual refusal bot messages

Pouyanpi opened the pull request:
## Description

Detect user input language and return refusal messages in the same language when content safety rails block unsafe content. Supports 9 languages: English, Spanish, Chinese, German, French, Hindi, Japanese, Arabic, and Thai.

## Language Detection Benchmark Results

### Datasets Used

| Dataset  | Description                             | Samples | Languages         |
|----------|-----------------------------------------|---------|-------------------|
| papluca  | [language-identification ](https://huggingface.co/datasets/papluca/language-identification#additional-information)    | 40,500  | 9 (all supported) |
| nemotron | [NVIDIA Nemotron-Safety-Guard-Dataset-v3](https://huggingface.co/datasets/nvidia/Nemotron-Safety-Guard-Dataset-v3) | 336,283 | 8 (missing zh*)   |

  Chinese samples in Nemotron are all REDACTED; Chinese coverage validated via papluca dataset.

#### Prompt Length Analysis (characters)

| Dataset  | Min | Max    | Mean  | P25 | P50 | P75 | P90 | P95   | P99   |
|----------|-----|--------|-------|-----|-----|-----|-----|-------|-------|
| papluca  | 2   | 3,657  | 129.8 | 50  | 97  | 162 | 258 | 351   | 627   |
| nemotron | 1   | 20,750 | 303.5 | 51  | 111 | 331 | 625 | 1,072 | 3,004 |

Note: fast-langdetect truncates input at 80 characters by default (`max_input_length=80`), so longer prompts are effectively evaluated on their first 80 chars.

---
### Overall Accuracy comparison

| Dataset  | Samples | fast-langdetect | lingua | detect_language action |
|----------|---------|-----------------|--------|------------------------|
| papluca  | 40,500  | 99.71%          | 99.79% | 99.71%                 |
| nemotron | 336,283 | 99.35%          | 99.46% | 99.42%                 |

---
### Latency comparison (μs)

| Dataset  | fast-langdetect Avg | fast-langdetect P95 | lingua Avg | lingua P95 | Action Avg | Action P95 |
|----------|---------------------|---------------------|------------|------------|------------|------------|
| papluca  | 12.12               | 15.54               | 116.21     | 205.29     | 25.77      | 28.75      |
| nemotron | 11.53               | 15.50               | 162.59     | 377.92     | 26.25      | 28.71      |

---
### Per Language Accuracy ([fast-langdetect](https://github.com/LlmKira/fast-langdetect/tree/main))

| Language      | papluca | nemotron |
|---------------|---------|----------|
| ar (Arabic)   | 98.87%  | 99.63%   |
| de (German)   | 99.93%  | 99.39%   |
| en (English)  | 100.00% | 99.03%   |
| es (Spanish)  | 100.00% | 99.04%   |
| fr (French)   | 99.98%  | 99.25%   |
| hi (Hindi)    | 98.76%  | 99.60%   |
| ja (Japanese) | 100.00% | 99.61%   |
| th (Thai)     | 99.93%  | 99.29%   |
| zh (Chinese)  | 99.93%  | N/A      |

---
### Per-Language Accuracy ([lingua](https://github.com/pemistahl/lingua-py))

| Language      | papluca | nemotron |
|---------------|---------|----------|
| ar (Arabic)   | 99.84%  | 99.75%   |
| de (German)   | 100.00% | 99.55%   |
| en (English)  | 99.93%  | 99.00%   |
| es (Spanish)  | 99.98%  | 99.43%   |
| fr (French)   | 99.82%  | 99.35%   |
| hi (Hindi)    | 98.80%  | 99.81%   |
| ja (Japanese) | 100.00% | 99.69%   |
| th (Thai)     | 99.78%  | 99.12%   |
| zh (Chinese)  | 99.93%  | N/A      |

---
**Why fast-langdetect?**

https://github.com/LlmKira/fast-langdetect

0. MIT license and [Creative Commons Attribution-Share-Alike License 3.0](https://github.com/LlmKira/fast-langdetect/blob/main/src/fast_langdetect/resources/NOTICE.MD).
1. comparable accuracy: within 0.1-0.5% of lingua across all datasets (99.35% vs 99.46% on 336k samples)
2. 10-14x faster: average latency ~12μs vs ~140μs
3. simpler integration: single lightweight dependency
4. no cold start issues: unlike lingua which requires model building
5. [no dependency issue in future](https://github.com/LlmKira/fast-langdetect/blob/main/pyproject.toml) 

### Error analysis

Most errors occur with:
- short text (single words): insufficient context for detection
- mixed language content: text containing English within non-English context
- similar language confusion: Spanish vs Galician, Hindi vs Marathi, Arabic vs Persian

The action correctly falls back to English (en) for unsupported detected languages.

---
### Benchmark Scripts
checkout to [temp/lang-detect-benchmark](https://github.com/NVIDIA-NeMo/Guardrails/tree/temp/lang-detect-benchmark) branch

Located in eval/language_detection/:
make sure to have datasets and pandas installed:

```
poetry run pip install pandas datasets
```

```sh
# run all benchmarks
poetry run python eval/language_detection/run_benchmarks.py

# Or run individually
poetry run python eval/language_detection/benchmark.py --dataset papluca --mode action --report eval/language_detection/reports/
```

Pouyanpi committed:
feat(content_safety): add auto selected multilingual refusal bot message support

Detect user input language and return refusal messages in the same
language when content safety rails block unsafe content. Supports 9
languages: English, Spanish, Chinese, German, French, Hindi, Japanese,
Arabic, and Thai.

Pouyanpi committed:
add tests

fix

Pouyanpi committed:
fix

cparisien commented on nemoguardrails/library/content_safety/actions.py:
```diff
@@ -220,3 +220,79 @@ async def content_safety_check_output(
+
+SUPPORTED_LANGUAGES: FrozenSet[str] = frozenset({"en", "es", "zh", "de", "fr", "hi", "ja", "ar", "th"})
+
+DEFAULT_REFUSAL_MESSAGES: Dict[str, str] = {
+    "en": "I'm sorry, I can't respond to that.",
+    "es": "Lo siento, no puedo responder a eso.",
+    "zh": "抱歉，我无法回应。",
+    "de": "Es tut mir leid, darauf kann ich nicht antworten.",
+    "fr": "Je suis désolé, je ne peux pas répondre à cela.",
+    "hi": "मुझे खेद है, मैं इसका जवाब नहीं दे सकता।",
+    "ja": "申し訳ありませんが、それには回答できません。",
+    "ar": "عذراً، لا أستطيع الرد على ذلك.",
+    "th": "ขออภัย ฉันไม่สามารถตอบได้",
+}
```
If we later had other multilingual rails, would we be repeating this mechanism in each rail? Or just the set of supported languages per rail? I don't think we need to do it now (since we don't have other multilingual rails to test it), but we should be aware of what refactoring would be needed to move the below language detection to a shared level.

Pouyanpi replied:
Of course, we can relax this constraint later and allow users more flexibility. Once we need to support other models or other types of rails (beyond content safety) that require multilingual responses, we can:

1. Move the `detect_language` action from `library/content_safety/actions.py` to a shared location (`nemoguardrails/actions/`) making it available to all rails
2. It was possible to also introduce a Colang level abstraction like `bot refuse to respond $multilang=true`, could be done easily for Colang 2.0, but I think it is better if we don't add new Colang features for now.


I agree, for now, keeping it scoped to content safety keeps the implementation focused.

cparisien commented on nemoguardrails/library/content_safety/actions.py:
```diff
@@ -220,3 +220,79 @@ async def content_safety_check_output(
+    try:
+        from fast_langdetect import detect
+
+        result = detect(text, k=1)
```
Does fast-langdetect ever return a full locale with dialect, like en-US versus en? I don't see it in the docs, but I do see some upper/lowercase inconsistency.

Pouyanpi replied:
fair point, thanks for raising it. Just took a closer look: the fast-langdetect source code and fastText model behavior.

- fast-langdetect [README](https://github.com/LlmKira/fast-langdetect/tree/main?tab=readme-ov-file#language-codes--english-names) mentions BCP-47 tags like `"zh-cn"`, `"pt-br"`
- but the [fastText lid.176.bin model](https://fasttext.cc/docs/en/language-identification.html) uses simple ISO 639 codes: `zh`, `pt`, `en`, etc.
- fast-langdetect source simply strips `__label__` prefix from fastText output,  no regional mapping is applied

**validated with actual test**:
```python
>>> detect("抱歉，我无法处理该请求", k=2)
[{'lang': 'zh', 'score': 0.80}, {'lang': 'ta', 'score': 0.08}]
```
returns `"zh"`, NOT `"zh-cn"`.

So no regional variant handling needed.

tgasser-nv commented:
This looks really good @Pouyanpi ! I have a few comments:

* Could you commit the evaluation scripts as well in the final PR (for reproducibility?)
* What does the "Action" column in the latency report refer to? Is this the latency end-to-end when fast-langdetect  embedded in a Guardrails action? It approximately doubles the mean and p95.
~~* Is it possible to customize refusal texts in Colang-only, or does it need a Python change?~~ Just saw this is in the RailsConfig, that's perfect.
* Could you calculate percentiles of prompt-length (ideally in tokens but characters is fine too) for each of the datasets?

Not needed in this PR, but I'm thinking of RAG prompts where we have LLM instructions, user query, and relevant context chunks are all in a flattened prompt. These prompts can be pretty long (up to 7k tokens in some cases). This isn't needed for  this PR, but I would be interested in a follow-on where we sample part of a prompt before running classification on the sample (e.g. 200 chars). This would be an optional config field. Customers would then have a knob to trade off accuracy vs latency for language detection.

Pouyanpi commented:
> * Could you commit the evaluation scripts as well in the final PR (for reproducibility?)

I've included them [temp/lang-detect-benchmark](https://github.com/NVIDIA-NeMo/Guardrails/tree/temp/lang-detect-benchmark) branch to make review easier. If you find it easier I will do.
But we don't intend to merge those in develop, right?

> * What does the "Action" column in the latency report refer to? Is this the latency end-to-end when fast-langdetect  embedded in a Guardrails action? It approximately doubles the mean and p95.

Yes  

>   ~* Is it possible to customize refusal texts in Colang-only, or does it need a Python change?~ Just saw this is in the RailsConfig, that's perfect.

Yes, I would like to avoid adding colang level features as much as possible

> * Could you calculate percentiles of prompt-length (ideally in tokens but characters is fine too) for each of the datasets?

Done! updated the description.

> Not needed in this PR, but I'm thinking of RAG prompts where we have LLM instructions, user query, and relevant context chunks are all in a flattened prompt. These prompts can be pretty long (up to 7k tokens in some cases). This isn't needed for this PR, but I would be interested in a follow-on where we sample part of a prompt before running classification on the sample (e.g. 200 chars). This would be an optional config field. Customers would then have a knob to trade off accuracy vs latency for language detection.

fast-langdetect already does the truncation by [default](https://github.com/LlmKira/fast-langdetect/tree/main?tab=readme-ov-file#input-handling) but indeed we can give that flexibility to the users :
- `max_input_length=80` characters (configurable)

tgasser-nv commented:
> > * Could you commit the evaluation scripts as well in the final PR (for reproducibility?)
> 
> I've included them [temp/lang-detect-benchmark](https://github.com/NVIDIA-NeMo/Guardrails/tree/temp/lang-detect-benchmark) branch to make review easier. If you find it easier I will do. But we don't intend to merge those in develop, right?
> 

Why wouldn't we merge them into develop? It's best practice in ML to make any results reproducible, for which we need the input datasets and scripts. The datasets are public and linked above. I'd imagine we'll have to re-run evals for new languages as they're added to the content-safety and other models. So we'll run this script periodically.

> > * What does the "Action" column in the latency report refer to? Is this the latency end-to-end when fast-langdetect  embedded in a Guardrails action? It approximately doubles the mean and p95.
> 
> Yes
> 
Was that measured at a concurrency of 1? Having a 100% overhead for each language inference is a lot higher than I'd expect. We don't need to fix it in this PR. 

> > ~* Is it possible to customize refusal texts in Colang-only, or does it need a Python change?~ Just saw this is in the RailsConfig, that's perfect.
> 
> Yes, I would like to avoid adding colang level features as much as possible
> 
**+1**

> > * Could you calculate percentiles of prompt-length (ideally in tokens but characters is fine too) for each of the datasets?
> 
> Done! updated the description.
> 
Could you check? I didn't see any length description. 

> > Not needed in this PR, but I'm thinking of RAG prompts where we have LLM instructions, user query, and relevant context chunks are all in a flattened prompt. These prompts can be pretty long (up to 7k tokens in some cases). This isn't needed for this PR, but I would be interested in a follow-on where we sample part of a prompt before running classification on the sample (e.g. 200 chars). This would be an optional config field. Customers would then have a knob to trade off accuracy vs latency for language detection.
> 
> fast-langdetect already does the truncation by [default](https://github.com/LlmKira/fast-langdetect/tree/main?tab=readme-ov-file#input-handling) but indeed we can give that flexibility to the users :
> 
> * `max_input_length=80` characters (configurable)

Could you add optional Pydantic fields for any of these values it makes sense to expose to users? Looking at the [config](https://github.com/LlmKira/fast-langdetect/blob/803a858a5b2615934b7a1f890f47b7ed74a3ea64/src/fast_langdetect/infer.py#L202-L210) I think `normalize_input`, `max_input_length`, and `model` are all fields users might care about

Pouyanpi committed:
feat(multilingual): add language detection configuration options

Add configurable parameters for language detection:
- max_text_length: Control maximum input text length for detection
- normalize_text: Toggle text normalization before detection
- cache_dir: Specify custom cache directory for detection models

Updated MultilingualConfig with new optional fields and modified
_detect_language to use LangDetector with custom configuration
instead of the simple detect function.

Pouyanpi committed:
update poetry lock

Pouyanpi commented:
Thanks @tgasser-nv! Yes you're right, here is how it looks (didn't save the change):

<img width="815" height="212" alt="image" src="https://github.com/user-attachments/assets/e2d5d470-d0e7-4d0d-8bc4-b8ce263559bd" />

I've extended the configuration but I'm not sure if we actually need to do it (I think better to revert its [commit](https://github.com/NVIDIA-NeMo/Guardrails/pull/1530/commits/4339ac2e73e1bd30118f0cf7d6433d719f0fe8a0)).
most users will just do `multilingual.enabled: true` and use library defaults. The config options (max_text_length, normalize_text, cache_dir) are edge cases.

Do we actually need these config options exposed?

  - cache_dir -> users can set FTLANG_CACHE env var
  - normalize_text -> library default (true) is almost always correct
  - max_text_length -> library default (80) is optimized for accuracy

  Maybe the cleanest solution is: don't expose these options at all and keep _detect_language(text) simple.

  If a power user really needs custom settings, they can:
  1. Use FTLANG_CACHE env var for cache location
  2. Or we add config options later when there's a real need

the `model` field looks library specific and we are better of not including it in the config.  I found cache_dir might be interesting but we can document FTLANG_CACHE. Do you think we should keep the new config options or revert?



Regarding the evaluation scripts and datasets: let's address that in a follow-up PR to keep this one focused. 
I think it's better to keep this type of analysis work in dedicated branches for now.

We can follow up in separate PRs once we establish a clear pattern for where these should live (e.g., scripts/benchmarks/, eval/, etc.) and how they should be maintained.

Pouyanpi committed:
chore(deps): add fast-langdetect dependency to dev

tgasser-nv commented:
> Thanks @tgasser-nv! Yes you're right, here is how it looks (didn't save the change):
> 
> <img alt="image" width="815" height="212" src="https://private-user-images.githubusercontent.com/13303554/524984628-e2d5d470-d0e7-4d0d-8bc4-b8ce263559bd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjU4MTUzMTcsIm5iZiI6MTc2NTgxNTAxNywicGF0aCI6Ii8xMzMwMzU1NC81MjQ5ODQ2MjgtZTJkNWQ0NzAtZDBlNy00ZDBkLThiYzQtYjhjZTI2MzU1OWJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE1VDE2MTAxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjNGM0N2UzMGIzMTgxYTUzNWQ5YzQwMDVlMGU5MzMzN2ZkNjVlNDgzNDI5Mjg1Y2JlNTIzNmFiNTVhYTNlZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tnhpDHNQpzbpFq5cCLHxfVKP0ZFpOKjglyKPvGFlVoo">
> I've extended the configuration but I'm not sure if we actually need to do it (I think better to revert its [commit](https://github.com/NVIDIA-NeMo/Guardrails/pull/1530/commits/4339ac2e73e1bd30118f0cf7d6433d719f0fe8a0)). most users will just do `multilingual.enabled: true` and use library defaults. The config options (max_text_length, normalize_text, cache_dir) are edge cases.
> 
> Do we actually need these config options exposed?
> 
> * cache_dir -> users can set FTLANG_CACHE env var
> * normalize_text -> library default (true) is almost always correct
> * max_text_length -> library default (80) is optimized for accuracy
> 
> Maybe the cleanest solution is: don't expose these options at all and keep _detect_language(text) simple.
> 
> If a power user really needs custom settings, they can:
> 
> 1. Use FTLANG_CACHE env var for cache location
> 2. Or we add config options later when there's a real need
> 
> the `model` field looks library specific and we are better of not including it in the config. I found cache_dir might be interesting but we can document FTLANG_CACHE. Do you think we should keep the new config options or revert?
> 
> Regarding the evaluation scripts and datasets: let's address that in a follow-up PR to keep this one focused. I think it's better to keep this type of analysis work in dedicated branches for now.
> 
> We can follow up in separate PRs once we establish a clear pattern for where these should live (e.g., scripts/benchmarks/, eval/, etc.) and how they should be maintained.

Thanks for explaining, I think you can revert the changes for these fields. We benchmarked performance and ran evals with these default settings for  `normalize_text` and `max_text_length` fields so we know it'll work well for our customers. Assuming their use-case matches the dataset domain. 

The `cache_dir` is a little trickier, I think using the FTLANG_CACHE env var is fine. The only case they'd need this is if they don't have write-permissions to a given directory.

tgasser-nv reviewed (approved):
Looks good, can you revert the [commit to add extra config fields](https://github.com/NVIDIA-NeMo/Guardrails/pull/1530/commits/4339ac2e73e1bd30118f0cf7d6433d719f0fe8a0)? I agree these wouldn't be needed for the majority of users. And the `cache_dir` can be set in an env var if they need a different writable directory in which to download the models.

Pouyanpi committed:
Revert "feat(multilingual): add language detection configuration options"

This reverts commit 4339ac2e73e1bd30118f0cf7d6433d719f0fe8a0.
